### ISSUE TITLE
Packaged Python cannot import binary modules in debug

### ISSUE THREAD

**Issue by evanchia-ly-sdets** [o3de/o3de#4404]
**Describe the bug**
The debug version of O3DE Editor uses the debug version of python. Currently, we do not include any debug versions of python packages that contain binary files (i.e. psutil, numpy, scipy, etc.). This means that when we attempt to import these modules in debug mode, an import error will occur.
Note: This only occurs on Windows

**To Reproduce**
1. Go to ~\o3de\python\runtime\python-3.7.10-rev2-windows\python
2. Open a command line at the directory
3. Invoke python_d.exe
4. Attempt to import a module that includes binaries (i.e. psutil)
5. The user receives an import error

**Expected behavior**
The user does not receive any import errors

**Desktop/Device (please complete the following information):**
 - Device: PC
 - OS: Windows

**Comment by Kadino** [comment]
It may be best for third party dependencies to use non-debug builds when O3DE builds in debug. I believe an RFC on this is necessary.

**Comment by amznestebanpapp** [comment]
I think @lawsonamzn has a RFC in the works for this
We have to tweak all 3rdparty to be able to use profile/release 3rdParty in our O3DE debug builds (for Windows). For other platforms it should work, but will require to regenerate all 3rdParty for all platforms so they are all consistent and only have one configuration in them.
This will mean that we will loose the ability to debug 3rdParty, or that we would need to do manual steps to use a 3rdParty debug version.

**Comment by lawsonamzn** [comment]
Yes, the essential problem is that debug and release python requires corresponding debug and release version of all binary modules.   But while the PyPI (pip install) repositories contain pre-built binaries for release python, very rarely do they contain debug.  
So to get debug working you'd have to build all the requirements.txt dependnecies from source, in debug mode.  This is _theoretically_ possible but it turns out each such dependency has its own build requirements, compiler requirements, etc, including some esoteric ones like a working COBOL environment for things like numpy.

Ultimately, the longer term plan (IMO, this is up for discussion still) is to use release python even in debug mode (since we aren't really interested in debugging python itself anyway).  On windows though, this would require us to also use Release Qt, and Release PySide (QtForPython) since they all have to be binary stable and linked to the same CRT runtime on windows and depend on each other.  

This would mean redoing many of our 3p libraries, and setting ITERATOR_DEBUG_LEVEL=0 on the entire engine, (at least, parts of it which link to 3ps, which is all parts) which would also make it so that debug 3p cannot be used at all.

So pros and cons here...

**Comment by moraaar** [comment]
>This will mean that we will loose the ability to debug 3rdParty, or that we would need to do manual steps to use a 3rdParty debug version.

Instead of losing the ability to use debug in all 3rdParty with o3de debug configuration, I'd suggest to use cmake macro so it could be decided what configuration to use in o3de debug. By default it will use debug, but if you set the macro to true `LY_O3DE_DEBUG_USE_RELEASE_3RDPARTY_LIBS=TRUE` then all the 3rdparty `FindX.cmake` will know that with debug they have to link with release version of the library.

PhysX already supports something similar to make o3de profile to use physx sdk checked version of the lib, instead of profile version of the lib. https://github.com/o3de/3p-package-source/blob/0fe74f4cbc5e4f2e8ce1fda43645474249df60d3/package-system/PhysX/FindPhysX.cmake.template#L26

**Comment by lawsonamzn** [comment]
Yes, we'd try to leave an option in to change it, it jjust wouldn't be the default (IMO).

Note that the only problem I'm aware of is ITERATOR_DEBUG_LEVEL at the ABI level, and only on Windows.   The CRT debug memory thing is no longer an issue.  

ITERATOR_DEBUG_LEVEL only affects CPP code, not C, so any 3p (including python itself, which AFAIK is a pure C library) can already just use release.  So really this revolves around things written in C++ which depend on python - for example, PySide2 (AKA QtForPython), which then in turn depends on Qt, and Pybind11, but that ones header only

**Comment by lawsonamzn** [comment]
I guess if we want a summary of what I recommend in terms of work here to enable this:
* Rework QtForPython, Qt, Python packages to always use release by default, unless some option is set.  These have to be done 'as a group' as they are all inter-dependent.
*

**Comment by moraaar** [comment]
Let me play devil's advocate here.

If this is just to get automated tests running on debug, we have to think what we gain by having that. Is it that important to run automated tests in debug? It sounds like it's a huge impact having to rebuild all 3rdParty and potentially losing using 3rdParty on debug. The option of reworking QtForPython seems more manageable, but still you mention is an intensive task. 

Also, running tests, editor and AP in debug is going can be extremely slow, it's going to need to run very long jobs in Jenkins, with the computational cost it implies.

**Comment by lawsonamzn** [comment]
Well the huge time investment is one reason this hasn't begun yet.  However, there are other benefits to using mostly release 3p, mainly in terms of being able to run debug code fairly quickly.  We actually already do this for many 3p due to how slow debug can be (for example, image processing).  Most of the time you don't want to actually be debugging 3p libraries (you still would get symbols and stuff in debug, just would be an optimized build).  Most of the time the code you are in is your own code.

Shippiing only release could also be a drastic size reduction for download.  especially if coupled with a symbol server instead of shipped-in-package PDBs, or sidecar packages containing symbols that are optional.

**Comment by byrcolin** [comment]
We should see if this happens on 3.10

**Comment by lawsonamzn** [comment]
FWIW if you absolutely need the binary debug moduels, you can definitely build them locally.
For example if you try to launch a python script in the _debug_ editor, and it tries to use a binary module like psutil, it wont be able to load it.  But what you CAN do is compile it using pip, locally

From the O3DE folder:
```
.\python\runtime\python-3.10.5-rev1-windows\python\python_d -m pip install --no-binary :all: --global-option build --global-option --debug --force-reinstall psutil==5.8.0 --ignore-installed
```
Note that you must use the actual python_d.exe to do this, you can't use the python.cmd as that will invoke the release version.

Once you've done the above command line, psutil_d will be installed side by side with psutil (non-d) and both will work.  However, you'd have to do this for every missing binary.

**Comment by spham-amzn** [comment]
What I would propose is to not run debug python by default when running debug O3DE. Debug python is only really needed if we are debugging python packages, but if we want to have python import any debug version of native C++ code, then we can enable using python debug through a cmake variable. The caviat here is that anyone who chooses to do so will also need to manually build all of the current modules from requirements.txt in debug as well.

**Comment by lawsonamzn** [comment]
I think I looked at this and the problem (at least with prior versions o python?) was that the debug and release python were not actually binary compatible in their ABI, especially when it comes to modules that linked to it.

   So using the release python would force us to use the release version of everthing that linked to python or used its libs (so, Qt/Pyside, assimp, opencolorio/openimageio, and the list goes on and on).  And the problem with that is that is that some of those are C++ libraries which have different iterator impls between debug and release (on windows). 

This is the main tangle of the problem that I encountered when I investigated doing 'release python on debug o3de'. 

But I could be wrong - more investigation needs to be done.

**Comment by spham-amzn** [comment]
Well, when you install python on windows, the setup gives you the option to also download debugging symbols and such as well. An alternative could be either document how to build those native modules locally in debug (as per @lawsonamzn suggestion), or have a script/method to do so based on the existing requirements.txt.

**Comment by lawsonamzn** [comment]
The issue we ran into while trying to build some of the other modules in debug is that the requirements for some of them is kinda wild, like, for example, having a working prolog compiler for numpy.
